### PR TITLE
Add emphasis to selected annotations

### PR DESCRIPTION
**Pull request by csillag**
### Current behavior:

Currently, we have the following methods for selecting a specific set of annotations:
 - Click on the highlight (when always-on mode is on)
 - Click on the bar on the bucket bar.

When a set of annotations is selected, their cards will be visible in the sidebar,
and everything else will be hidden. The contents of the sidebar doesn't change,
even if the user scrolls the document.

### Proposed change:

This PR adds the following extra behavior:

The highlights belonging to the selected annotations stays visible in the document,
even if always-on mode is off, and even if the pointer is not currently hovering
over the annotation card.

(As described in #800 )

### Todo:

 - The styling is not intended to be final. Someone who cares about colors should
   devise the best color scheme, taking into account all highlight states.
   (focus, always-on mode, and selection.)
 - Maybe this could be extended to search, too.
   (Currently, search is not impacted at all.)

  * * *

The tests have been updated, and new tests have been added to cover the new functionality.

### CODE REVIEW

File: h/static/scripts/controllers.coffee [hunk 1]
```diff
@@ -118,6 +121,9 @@ class AppController
     $scope.clearSelection = ->
       $scope.search.query = ''
       annotationUI.clearSelectedAnnotations()
+      crossframe.notify
```
Comment (nickstenning) [inline h/static/scripts/controllers.coffee]: Something about this seems wrong to me. Specifically, the `annotationUI` service is supposed to be the canonical source of the annotation UI state. That means that we shouldn't have to call two different methods to call in order to clear the selection in both frames.

Perhaps the right thing to do is to move the `annotationUI` mutator method into a listener in `annotationUISync`, so that only the crossframe call is here.

Comment (csillag) [inline h/static/scripts/controllers.coffee]: I'll check that out.

**Comment by dwhly** [comment]
I like this change, with the exception that I'm struggling w the introduction of a new color (green) for highlights when cards are selected.  I see what you're trying to do, and I appreciate the distinction that the additional color gives, particularly when there are a bunch of annotations.

However, I'm fairly sure we're probably going to introduce the ability to change highlight colors at some point, since it seems to be a thing folks want-- probably best not to keep overloading colors with meaning.  

Another option (not sure if it makes the code tricky) might be to use [$highlight-color-third](https://github.com/hypothesis/h/blob/master/h/static/styles/variables.scss#L59) instead.  I imagine that if we implement highlight colors, that instead of having hard coded overlapping color values, we'd instead use a [blending](https://css-tricks.com/basics-css-blend-modes/) approach (?) that would apply regardless of color.  So, the approach here would be to go to the darkest overlapping value when something is selected.

**Comment by csillag** [comment]
> I'm struggling w the introduction of a new color (green) for highlights when cards are selected.

That color was never meant to stick. As I wrote in the OP:

> The styling is not intended to be final. 

I intentionally made it totally different from the existing colors so that you all (= the other team members) can discern the functionality more clearly, when discussing this PR.

**Comment by nickstenning** [comment]
It would be good to close this out one way or the other.

1. Have you looked into my earlier comments?
2. Could you coordinate with whoever you need to in order to decide on the styling you want to merge. I'm not inclined to merge this with "temporary" styling.

**Comment by csillag** [comment]
> It would be good to close this out one way or the other.

Sure - but this PR is only 3 days old, 2 of which was a weekend, so I don't think we are terribly slow on this.

> 1. Have you looked into my earlier comments?

I did, and I agree with them. I will do what you suggested after finishing a bunch of higher priority stuff.

> 2. Could you coordinate with whoever you need to in order to decide on the styling you want to merge. I'm not inclined to merge this with "temporary" styling.

@dwhly already suggested something; I'll simply do that, unless someone else proposes something else.

**Comment by tilgovi** [comment]
What's the purpose of this pull request? Should we not simply hide highlights that aren't selected?

**Comment by csillag** [comment]
> What's the purpose of this pull request?

As described in the OP:

> The highlights belonging to the selected annotations stays visible in the document, even if always-on mode is off, and even if the pointer is not currently hovering over the annotation card.

> Should we not simply hide highlights that aren't selected?

Highlights _are_ hidden, when they aren't selected.
The problem that this PR tries to solve is then even when an annotation is selected, the highlights are still hidden by default, unless always-on mode is on, or the pointer is hovering above the card.

This PR changes that by making the highlights belonging to the selected cards *always* visible.

You can check out the dokku for live demo.

**Comment by tilgovi** [comment]
I understand what you're saying, but I don't think you understand me.

I'm not debating the functionality, but the mechanism. It's not clear to me that we need more than one highlight class. Things are either highlight or they are not and if we haven't the means to toggle always on/off and selection in a way where the state is reflected in the annotation(s) having _the_ highlight class then we have not factored our code well.

**Comment by csillag** [comment]
The confusion stems from the fact that we are using the word *highlight* with several different meanings, in the same discourse:
 * First, it denotes the things that we have in the DOM
 * Second, it also denotes a *state* or *activity*, when these things are visible.

I propose that we find a better work for one of them. Maybe let's call the second one *shown* or *activated* ?

**Comment by tilgovi** [comment]
One is a target (or, in your preferred language, an anchor). A highlight is highlighted.

**Comment by csillag** [comment]
> One is a target (or, in your preferred language, an anchor). A highlight is highlighted.

That's not how we use those terms in our code.
 * A `target` is what we store in the annotation
 * An `anchor` is the exact position we calculated for a target, given the current state of the document.
 * A `highlight` is the manifestation of an `anchor` in the DOM.

We call the SPANS that we add highlighs, even when they are invisible. The names of our CSS classes are: `annotation-hl` for all highlights (including invisible), and `annotation-hl-activated` for the activated (ie. visible) ones.

**Comment by tilgovi** [comment]
From now on I'm not going to call something a highlight unless it's highlighted. A span is just a target. If you prefer, call it an anchor. I don't think there's a useful distinction, except to say that one is the serialized data and the other is in-memory in the document model.

**Comment by tilgovi** [comment]
But, to get back to the _point_:

- I would love to reconsider the highlight visibility toggle
- I would be interested in only setting the highlight class on things which are highlighted, and removing the class if we wish them to be "deselected" or otherwise not shown.

**Comment by csillag** [comment]
> From now on I'm not going to call something a highlight unless it's highlighted.

Whatever expressions we decide to use, we should use them consistently in our code, UI and discussions consistently. Right now, we consistently call those highlights, even when they are invisible.

If you think we should stop calling something a highlight unless it's highlighted (which certainly makes sense, if you ask me), then
  * we should find a new name for them, and 
  * submit a PR that replaces all occurrences in the code, and
  * then we should change the words that we use in our discussions.

I consider changing the wording discussions, while leaving the code as it is, to be seriously misleading.

**Comment by csillag** [comment]
> I would be interested in only setting the highlight class on things which are highlighted, and removing the class if we wish them to be "deselected" or otherwise not shown.

I think we use that class for triggering various actions. So instead of removing it, maybe we should rename it to something that does not involve the word `highlight`. (But let's avoid the words that are already used for other things, including `target`, `anchor` and `annotation`.)

**Comment by tilgovi** [comment]
Anyway, to get back to the matter at hand: how should we close this out? Do we like the selected state thing that's happening here?

**Comment by dwhly** [comment]
I like it-- with the change noted to avoid the third color.

**Comment by RawKStar77** [comment]
I think it's confusing. I don't quite know whats causing this behavior. I think this adds additional complexity and not sure why it's extremely useful. What is the user story behind this?

**Comment by dwhly** [comment]
> I don't quite know whats causing this behavior. 

Which behavior?

> I think this adds additional complexity and not sure why it's extremely useful. What is the user story behind this?

As a user, when annotations are selected, I want to see where they anchor in the text.  

Currently if annotations are selected in the sidebar, they're not highlighted in the text.  This fixes that.  Certainly any new code adds complexity.  But there's no new complexity from a UX perspective, since there are no new controls.  Only annotations whose highlights can be seen on the text when selected.

**Comment by gergely-ujvari** [comment]
:+1:  for having the selected state. It helps to notice more that he/she is looking at only a subset of the cards even if he/she scrolled down enough not to see the selection counter/clear selection parts on the top.

**Comment by RawKStar77** [comment]
I was speaking in terms of complexity *from a UX perspective.* 

Currently, if you focus an annotation or hover over it, it shows the corresponding highlight in the text. I think that model is good enough for the user story of a user wanting to see where *an annotation* anchors in a text. I guess what I don't get is: why would a user want to focus a group of annotations? 

More importantly, I think this as it currently operates, this PR actually introduces a bunch of unexpected behavior (putting aside the new green color). Why is the text now permanently highlighted even if always-on mode is off? Why do I now have to clear selection every time I want that highlight to go away?

I would be a big proponent of doing some user testing around this, and making tweaks before we merge this. 

Thoughts?

**Comment by RawKStar77** [comment]
Perhaps what I arguing we is that we should make this selected state shown only if the sidebar is open... perhaps that would be cleared. When I close the sidebar I've moved on to reading or highlighting the next thing, and I don't want to have to clear the selection to make the highlight go away.

**Comment by tilgovi** [comment]
Definitely at a minimum we need to clear the selection state or at least
unhighlight the selected targets.

Definitely still skeptical of all of this. I'm not a fan of the bucket
tabs, the selection, at all. What we're discussing here entrenches some of
these ideas by trying to refine how they present. I'd prefer not to have
them at all.

File: h/static/scripts/controllers.coffee (hunk 1 above)
Comment (csillag) [inline h/static/scripts/controllers.coffee]: > Something about this seems wrong to me. Specifically, the `annotationUI` service is supposed to be the canonical source of the annotation UI state. That means that we shouldn't have to call two different methods to call in order to clear the selection in both frames.

I agree that it would be nicer if we should only talk to a single component.

> Perhaps the right thing to do is to move the `annotationUI` mutator method into a listener in `annotationUISync`, so that only the crossframe call is here.

To make that work, we would have to teach the host side to echo these messages back to the sidebar. I would evaluate other approaches, too. What about teaching `AnnotatorUI` to transmit the selection info the the host side using the crossframe service?

Comment (csillag) [inline h/static/scripts/controllers.coffee]: @nickstenning, how about introducing a wrapper service (something like `AnnotationSelectionService`), which would provide a public API for defining/changing the selection, and having this intermediary service talk to both `AnnotatorUI`, and make the cross-frame calls?

**Comment by csillag** [comment]
> From now on I'm not going to call something a highlight unless it's highlighted. 

Apologies, but until we find a better name for our existing constructs, I'll call the things by their current name, in order to avoid confusion.

**Comment by csillag** [comment]
@RawKStar77:

> I think this adds additional complexity [...] 
> Currently, if you focus an annotation or hover over it, it shows the corresponding highlight in the text. I think that model is good enough for the user story of a user wanting to see where an annotation anchors in a text.

While studying the body (or replies) of the annotation, the user might want to look back to the text, and see where all this comes from. Currently, this requires using the mouse, but this should not be necessary.

> I guess what I don't get is: why would a user want to focus a group of annotations?

The ability to focus on a group of annotations does not come from this PR. You can focus *and* select a group of annotations on current master, too; just hover or click on a bucket.

> More importantly, I think this as it currently operates, this PR actually introduces a bunch of unexpected behavior (putting aside the new green color). Why is the text now permanently highlighted even if always-on mode is off?

This is not unexpected; this is *exactly* the single new feature added by this PR. The text is now permanently highlighted so that the user can always see where it is in the text.

> Why do I now have to clear selection every time I want that highlight to go away?

Because until you clear the selection, we have every right to assume that you are actively looking at the currently selected annotation(s), and so it's probably useful to see them in the text, too.

> Perhaps what I arguing we is that we should make this selected state shown only if the sidebar is open... perhaps that would be clearer. When I close the sidebar I've moved on to reading or highlighting the next thing, and I don't want to have to manually clear the selection to make the highlight go away.

Again, this is not something that is changed in this PR. On current master, the selection is retained, even if you close the sidebar and scroll away. If you have selected something, it will stay selected until you clear the selection, or select something else. So that's not new.

**Comment by csillag** [comment]
@nickstenning: I did the changes in the code that you proposed. (Now we don't send the same commands to two components differently; we only use the cross-frame service, and the annotationUI service is automatically synced to that.

**Comment by csillag** [comment]
@dwhly, I tweaked the color as you suggested. Dokku coming soon.

**Comment by csillag** [comment]
I think @dwhly has summarized the purpose of this PR well:

> As a user, when annotations are selected, I want to see where they anchor in the text.
>
> Currently if annotations are selected in the sidebar, they're not highlighted in the text. This fixes that. 

Really, it's not that complicated.

**Comment by csillag** [comment]
@tilgovi:

> I'm not a fan of the bucket tabs, the selection, at all.
> What we're discussing here entrenches some of these ideas
> by trying to refine how they present.
> I'd prefer not to have them at all.

I don't think what you say here is relevant to the current PR. If you would like to run an experiment with a completely different UI paradigm (ie. no bucket tabs, no selection), I think that's a great idea; go for it. I am interested to see how that turns out.

However, until we decide to migrate to an alternate UI, I fail to see why would it be a bad thing to refine how we use the current concepts. If this PR is an advancement *within our current paradigm*, then I think it's a good enough reason to merge it; even if we think that we might replace our current paradigm with something else in the future.

**Comment by csillag** [comment]
> Dokku coming soon.

https://selection.dokku.hypothes.is/

**Comment by csillag** [comment]
Oh yes. The tests are broken. I'll update them when we have settled on the code.

**Comment by tilgovi** [comment]
I like this.

File: h/static/scripts/controllers.coffee [hunk 2]
```diff
@@ -118,7 +118,7 @@ class AppController
 
     $scope.clearSelection = ->
       $scope.search.query = ''
-      annotationUI.clearSelectedAnnotations()
```
Comment (tilgovi) [inline h/static/scripts/controllers.coffee]: Seems like `selectAnnotations` should be on the annotation UI.

**Comment by tilgovi** [comment]
We should be leveraging annotation-ui-sync instead of interacting with the cross-frame service directly.

**Comment by tilgovi** [comment]
So, behavior is nice. I think we need to use annotation-ui-sync and annotation-ui for all this and makes ure our view controllers don't need to touch crossframe more than they already do.

Other than that, changing how we talk about highlights, etc, can happen later. For now, $highlight-color-third will work.

**Comment by csillag** [comment]
> Seems like `selectAnnotations` should be on the annotation UI.

It is; see [here](https://github.com/hypothesis/h/pull/2036/files#diff-476faf8b6d1f82c8a0a87197b314603dL53). The problem is that we can't call that directly; see below.

> We should be leveraging annotation-ui-sync instead of interacting with the cross-frame service directly.

We *are* leveraging annotation-ui-sync. We are using it to [set](https://github.com/hypothesis/h/blob/800-fancy-selected-annotations/h/static/scripts/annotation-ui-sync.coffee#L23) and [mutate](https://github.com/hypothesis/h/blob/800-fancy-selected-annotations/h/static/scripts/annotation-ui-sync.coffee#L29) the selection state on `annotationUI`, based on the information that comes in on the bridge.

That's why we don't have to talk to `annotationUI` directly in the controllers; only the `crossframe`.

> I think we need to use annotation-ui-sync and annotation-ui for all this and makes ure our view controllers don't need to touch crossframe more than they already do.

`annotation-ui-sync` depends on `annotation-ui`.
`crossframe` depends on `annotation-ui-sync` and `annotation-ui`.
Which means I can't make `annotation-ui` or `annotation-ui-sync` depend on `cross-frame`, because that would be a circular dependency. That's why ended up talking to the `cross-frame` in the controllers instead, and let `annotation-ui-sync` channel the selection towards `annotation-ui`. (Besides, @nickstenning suggested the same thing, too.)

What I could do instead of this is to create a different service, called `annotation-selection` or something similar, which would depend on `cross-frame`, and then in the controllers, we could talk to this `annotation-selection` service, which would talk to the `cross-frame`, and then (upon receiving the echo) the `annotation-ui-sync` service could talk to `annotation-ui`, as it already does.

Would you consider that an improvement?

**Comment by csillag** [comment]
@nickstenning, a while back, you wrote this:

> [...] the `annotationUI` service is supposed to be the canonical source of the annotation UI state. That means that we shouldn't have to call two different methods to call in order to clear the selection in both frames.
> 
> Perhaps the right thing to do is to move the `annotationUI` mutator method into a listener in `annotationUISync`, so that only the crossframe call is here.

This has been done since. How do you like it now?

**Comment by csillag** [comment]
@tilgovi, I also replied to your comments above; do you have any further observations?

**Comment by tilgovi** [comment]
I still need to look at this again. I thought a lot of effort was expended by aron and treora so that cross frame calls like these were handled automatically.

**Comment by csillag** [comment]
> I thought a lot of effort was expended by aron and treora so that cross frame calls like these were handled automatically.

Here is a short summary.

The `annotation-sync` service works both ways. (It synchronizes the state of annotations across both frames.) However, the `annotation-ui-sync` service is basically a one-way thing: it updates the sidebar's `annotation-ui` so that it stays in sync with the selection/focus state coming from the document side, and currently we don't have anything that would automatically update the same state on the document side, based on things happening in the sidebar.

In this PR, I am fully utilizing all the sync infrastructure already there: the `annotation-ui` is still updated by the `annotation-ui-sync` service, based on what's happening on the document side. The only manual cross-frame calls we need to make are the ones that carry information/commands from the sidebar to the document side. 

We had those manual cross-frame calls in the controllers to transmit the focus state, too; see [here](https://github.com/hypothesis/h/pull/2036/files#diff-ba740b145657584b68d9243979907788L228).

So, to summarize, I am using the existing sync infrastructure (built by @aron  and @Treora ) wherever it's present. Unfortunately, it doesn't yet do everything we need. Maybe we could extend it later ... but that's outside the scope of the current PR.

**Comment by tilgovi** [comment]
I can't tell how this is supposed to work. It looks like you define `getHighlights` as a static method on the `TextHighlight` class but then you call it on `annotator.anchoring`.

I would also very much appreciate it if you could take a stab at simplifying the language and the number of concepts here. It seems like we now have `getAnnotations`, `getHighlights` and `getAnchors`. All of these things, I think, have access to one another, so it seems a bit overkill to have all these things and probably confuses the responsibilities of components.

If anchoring is responsible for anchors it should only have `getAnchors`. It would be lovely if, as we discussed, we could start to view highlights as merely a style state of an anchor and not a separate object.

Whatever you can do to move in this direction would be appreciated, but at the least please clarify how this even works.

**Comment by csillag** [comment]
> I can't tell how this is supposed to work. It looks like you define `getHighlights` as a static method on the `TextHighlight` class but then you call it on `annotator.anchoring`.

Those are different methods. (The names are the same, but they are defined on different classes.)

 * [This one](https://github.com/hypothesis/h/pull/2036/files#diff-130344517827f6cc46967dbcc766d7b1R53) is used only [here](https://github.com/hypothesis/h/pull/2036/files#diff-402a7f217cd007dee77c51e27b85fedcR352). (The `event.data` attribute comes from [here](https://github.com/hypothesis/h/pull/2036/files#diff-130344517827f6cc46967dbcc766d7b1R73) - see the [docs](https://api.jquery.com/delegate/#delegate-selector-eventType-eventData-handler) for he jQuery magic involved.)
 * [This one](https://github.com/hypothesis/h/blob/800-fancy-selected-annotations/h/static/scripts/annotator/plugin/enhancedanchoring.coffee#L263), on the other hand, is used in the calls you saw happening on  `annotator.anchoring`.

Maybe we could rename the first one to `getImpactedHighlights()`, because it's for determining the highlights that a given mouse event impacted.

**Comment by csillag** [comment]
> I would also very much appreciate it if you could take a stab at simplifying the language and the number of concepts here. It seems like we now have `getAnnotations`, `getHighlights` and `getAnchors`. All of these things, I think, have access to one another, so it seems a bit overkill to have all these things

Again, we are talking about different APIs here.
 * There is the [Enhanced Anchoring plugin](https://github.com/hypothesis/h/blob/800-fancy-selected-annotations/h/static/scripts/annotator/plugin/enhancedanchoring.coffee#L129), commonly referenced as annotator.anchoring, which provides [getAnchors()](https://github.com/hypothesis/h/blob/800-fancy-selected-annotations/h/static/scripts/annotator/plugin/enhancedanchoring.coffee#L271) (for collecting the anchors) and [getHighlights()](https://github.com/hypothesis/h/blob/800-fancy-selected-annotations/h/static/scripts/annotator/plugin/enhancedanchoring.coffee#L263) (for collecting the corresponding highlights). We could do with only the first one; the second one is a convenience method, and it's result can be generated from the results of the first one.
 * There is the `eventData` object that we pass to the event handlers for the mouse events happening over the TextHighlight. This object has two relevant methods: [getHighlights()](https://github.com/hypothesis/h/pull/2036/files#diff-130344517827f6cc46967dbcc766d7b1R53) (for collecting the impacted highlight objects), and [getAnnotations()](https://github.com/hypothesis/h/pull/2036/files#diff-130344517827f6cc46967dbcc766d7b1R45) for collecting the underlying annotations. Again, the second one is a convenience method, because the data can be extracted from the result of the first one.

> and probably confuses the responsibilities of components. If anchoring is responsible for anchors it should only have `getAnchors`.

As per our current definitions, the process of "anchoring" has two phases. The first one identifies the location in the document (and thus creates the anchor), and the second phase creates the physical marker in the DOM (aka. highlight). So in this sense, it's also responsible for highlights. That's why it has a `getHighlights()` method, too.

> It would be lovely if, as we discussed, we could start to view highlights as merely a style state of an anchor and not a separate object.

I would love to continue that discussion, because it has been cut short, and I haven't yet stated my view. We are not yet within sight of an agreement on that point. However, I can say that I fully support renaming things in a way that makes more sense.

> Whatever you can do to move in this direction would be appreciated,

I propose that we do any renamings in a different PR. 

> but at the least please clarify how this even works.

I am happy to answer any further questions you might have.

**Comment by tilgovi** [comment]
Rebased.

**Comment by tilgovi** [comment]
Lost a test in the rebase. The test you had in controllers-test didn't make any sense to me. It claimed to be testing that the crossframe was notified to deselect on delete but I can't see any code path that does that.

**Comment by csillag** [comment]
> Rebased.

Thanks! Will check it out on Monday.

> Lost a test in the rebase. The test you had in controllers-test didn't make any sense to me. It claimed to be testing that the crossframe was notified to deselect on delete but I can't see any code path that does that.

That's because I haven't yet wrote the tests for the current version. Any tests you might have seen have been written for an earlier, different implementation. As I stated [earlier](https://github.com/hypothesis/h/pull/2036#issuecomment-85302198):

> The tests are broken. I'll update them when we have settled on the code.

**Comment by csillag** [comment]
@tilgovi on IRC:

> the whole way we're doing this UI sync business is crazy
> I feel like we should just have the AnnotationUI API and during the application configuration,
> if we're in an iframe we should decorate the service so sysnc [...] (transmission lost)

@tilgovi, in the context of the current PR, I am not trying to reinvent the way we are doing the UI sync between the two frames. I am just perusing the framework and patterns laid down by @aron and @Treora.

I suggest we limit the review of this PR to what this PR does, and discuss refactoring the whole UI sync system on the mailing list, independently of this PR. 

Could we do that, please?

**Comment by nickstenning** [comment]
Closing this for now.
